chunk/dynamo_table_manager.go

```diff
@@ -0,0 +1,347 @@
+	m.wait.Add(1)
+	go m.loop()
+}
+
+// Stop stops the DynamoTableManager
+func (m *DynamoTableManager) Stop() {
+	close(m.done)
+	m.wait.Wait()
+}
+
+func (m *DynamoTableManager) loop() {
+	defer m.wait.Done()
+
+	ticker := time.NewTicker(m.cfg.DynamoDBPollInterval)
+	defer ticker.Stop()
+
+	if err := instrument.TimeRequestHistogram(context.Background(), "DynamoTableManager.syncTables", syncTableDuration, func(ctx context.Context) error {
+		return m.syncTables(ctx)
+	}); err != nil {
+		log.Errorf("Error syncing tables: %v", err)
+	}
+
+	for {
+		select {
+		case <-ticker.C:
+			if err := instrument.TimeRequestHistogram(context.Background(), "DynamoTableManager.syncTables", syncTableDuration, func(ctx context.Context) error {
+				return m.syncTables(ctx)
+			}); err != nil {
+				log.Errorf("Error syncing tables: %v", err)
+			}
+		case <-m.done:
+			return
+		}
+	}
+}
+
+func (m *DynamoTableManager) syncTables(ctx context.Context) error {
+	expected := m.calculateExpectedTables(ctx)
+
+	toCreate, toCheckThroughput, err := m.partitionTables(ctx, expected)
+	if err != nil {
+		return err
+	}
+
+	if err := m.createTables(ctx, toCreate); err != nil {
+		return err
+	}
+
+	return m.updateTables(ctx, toCheckThroughput)
+}
+
+type tableDescription struct {
+	name             string
+	provisionedRead  int64
+	provisionedWrite int64
+}
+
+type byName []tableDescription
+
+func (a byName) Len() int           { return len(a) }
+func (a byName) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }
+func (a byName) Less(i, j int) bool { return a[i].name < a[j].name }
+
+func (m *DynamoTableManager) calculateExpectedTables(_ context.Context) []tableDescription {
+	if !m.cfg.UsePeriodicTables {
+		return []tableDescription{
+			{
+				name:             m.tableName,
+				provisionedRead:  m.cfg.ProvisionedReadThroughput,
+				provisionedWrite: m.cfg.ProvisionedWriteThroughput,
+			},
+		}
+	}
+
+	result := []tableDescription{}
+
+	var (
+		tablePeriodSecs = int64(m.cfg.TablePeriod / time.Second)
+		gracePeriodSecs = int64(m.cfg.CreationGracePeriod / time.Second)
+		maxChunkAgeSecs = int64(m.cfg.MaxChunkAge / time.Second)
+		firstTable      = m.cfg.PeriodicTableStartAt.Unix() / tablePeriodSecs
+		lastTable       = (mtime.Now().Unix() + gracePeriodSecs) / tablePeriodSecs
+		now             = mtime.Now().Unix()
+	)
+
+	// Add the legacy table
+	{
+		legacyTable := tableDescription{
+			name:             m.tableName,
+			provisionedRead:  m.cfg.ProvisionedReadThroughput,
+			provisionedWrite: minWriteCapacity,
+		}
+
+		// if we are before the switch to periodic table, we need to give this table write throughput
+		if now < (firstTable*tablePeriodSecs)+gracePeriodSecs+maxChunkAgeSecs {
+			legacyTable.provisionedWrite = m.cfg.ProvisionedWriteThroughput
+		}
+		result = append(result, legacyTable)
+	}
+
+	for i := firstTable; i <= lastTable; i++ {
+		table := tableDescription{
+			// Name construction needs to be consistent with chunk_store.bigBuckets
```

Comment:
Not 100% sure, and getting sure would probably require doing the refactoring myself.